lib/pure/asyncfutures.nim

```diff
@@ -25,6 +25,9 @@ type
 
   FutureBase* = ref object of RootObj  ## Untyped future.
     callbacks: CallbackList
+    thenCb: CallbackFunc # Callback for Future.then
+    catchCb: CallbackFunc # Callback for Future.catch
+    finalCb: CallbackFunc # Callback for Future.finally
```

Comment:
these 3 add overhead to all uses of future, even those not using these facilities (which, if you're using `await` like is expected, is the majority of code)

Reply:
Their default value is nil. The callbacks list is an overhead too, but it work well... Do you still think it would be an important overhead ? How can I improve it ?

Reply:
well, it also adds deallocation overhead because they need to be checked if they're nil - the better question however is, what's advantage over:
```nim
proc cb(fut: Future[T]) =
  try: await fut
  except:..
  finally:..
```

or indeed

```nim
try: 
  await myFuture
except:
finally:
```

Reply:
Sure, your approach is good too. But the then-catch-finally approach can help to write very simple and short code. 

- Example 1

```nim
fut.finally: sendNotif "finished"
```

- Example 2

```nim
fut.catch: echo "bg task failed !"
```

Reply:
finalCb is the same as `callbacks`, and the other two are `callbacks` + a if

See https://github.com/status-im/nim-chronos/pull/254/files for instance